File: packages/protocol/src/state.ts
/**
 * Agent state machine — represents the current state of the AI coding agent.
 */
export type AgentState =
  | "idle"
  | "thinking"
  | "coding"
  | "running_command"
  | "waiting_approval"
  | "error";

/**
 * Token usage information for a completed turn.
 */
export interface TokenUsage {
  inputTokens: number;
  outputTokens: number;
  cachedInputTokens?: number;
}

/**
 * A single file change reported by the agent.
 */
export interface FileChange {
  path: string;
  kind: "add" | "delete" | "update";
}

export interface DiffLine {
  kind: "context" | "add" | "delete";
  text: string;
}

export interface DiffHunk {
  oldStart: number;
  oldLineCount: number;
  newStart: number;
  newLineCount: number;
  lines: DiffLine[];
}

export interface DiffFile {
  path: string;
  kind: FileChange["kind"];
  addedLines?: number;
  deletedLines?: number;
  isTruncated: boolean;
  truncationReason?: string;
  totalHunkCount: number;
  loadedHunks: DiffHunk[];
  nextHunkIndex?: number;
}

/**
 * Information about a running or completed agent session.
 */
export interface SessionInfo {
  id: string;
  agentType: "codex" | "claude";
  workDir: string;
  state: AgentState;
  createdAt: number;
  lastActiveAt: number;
}

File: crates/codepilot-protocol/src/state.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentState {
    Idle,
    Thinking,
    Coding,
    RunningCommand,
    WaitingApproval,
    Error,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum AgentType {
    Codex,
    Claude,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum FileChangeKind {
    Add,
    Delete,
    Update,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TokenUsage {
    pub input_tokens: u64,
    pub output_tokens: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cached_input_tokens: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FileChange {
    pub path: String,
    pub kind: FileChangeKind,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum DiffLineKind {
    Context,
    Add,
    Delete,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DiffLine {
    pub kind: DiffLineKind,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DiffHunk {
    pub old_start: u64,
    pub old_line_count: u64,
    pub new_start: u64,
    pub new_line_count: u64,
    pub lines: Vec<DiffLine>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DiffFile {
    pub path: String,
    pub kind: FileChangeKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub added_lines: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub deleted_lines: Option<u64>,
    pub is_truncated: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub truncation_reason: Option<String>,
    pub total_hunk_count: u64,
    pub loaded_hunks: Vec<DiffHunk>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub next_hunk_index: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SessionInfo {
    pub id: String,
    pub agent_type: AgentType,
    pub work_dir: String,
    pub state: AgentState,
    pub created_at: i64,
    pub last_active_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FileSearchMatch {
    pub path: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub directory_hint: Option<String>,
}
